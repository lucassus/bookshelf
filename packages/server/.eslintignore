coverage
dist
node_modules
jest.config.js
src/**/*.generated.ts
