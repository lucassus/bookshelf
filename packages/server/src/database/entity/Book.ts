import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import { Author } from "./Author";

@Entity({ name: "books", orderBy: { title: "ASC" } })
export class Book {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  title: string;

  @Column({ name: "cover_path" })
  coverPath: string;

  @ManyToOne(() => Author, (author) => author.books)
  @JoinColumn({ name: "author_id" })
  author: Promise<Author>;

  @Column({ name: "author_id" })
  authorId: number;

  @Column({ default: false })
  favourite: boolean;

  @CreateDateColumn({ name: "created_at" })
  created_at: Date;

  @UpdateDateColumn({ name: "updated_at" })
  updated_at: Date;
}
