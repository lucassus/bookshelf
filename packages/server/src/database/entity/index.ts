export { Author } from "./Author";
export { Avatar } from "./Avatar";
export { Book } from "./Book";
export { BookCopy } from "./BookCopy";
export { User } from "./User";
