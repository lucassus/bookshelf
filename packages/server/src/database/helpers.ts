import { SecureId } from "./SecureId";

type EntityTypes = "Author" | "Book" | "User" | "BookCopy";

export const secureId = new SecureId<EntityTypes>({
  separator: "-"
});
