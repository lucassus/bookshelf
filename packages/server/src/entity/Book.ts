import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import { Author } from "./Author";

@Entity({ name: "books" })
export class Book {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  title: string;

  @Column({ name: "cover_path" })
  coverPath: string;

  @Column({ name: "author_id" })
  authorId: number;

  @ManyToOne(() => Author, (author) => author.books)
  @JoinColumn({ name: "author_id" })
  author: Promise<Author>;

  @CreateDateColumn({ name: "created_at" })
  created_at: Date;

  @UpdateDateColumn({ name: "updated_at" })
  updated_at: Date;
}
