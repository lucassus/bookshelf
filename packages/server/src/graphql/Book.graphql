type Book implements Resource {
  id: ID!
  title: String!
  description: String
  cover: Image!
  favourite: Boolean!
  author: Author
  copies: [BookCopy!]!
}
