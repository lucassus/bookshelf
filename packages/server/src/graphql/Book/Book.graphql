type Book implements Resource & Timestampable {
  id: ID!
  title: String!
  description: String
  cover: Image!
  favourite: Boolean!
  author: Author
  copies: [BookCopy!]!
  createdAt: String!
  updatedAt: String!
}
