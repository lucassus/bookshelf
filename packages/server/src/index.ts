import express from "express";
import path from "path";
import "reflect-metadata";

import { PORT } from "./config";
import { createConnection } from "./database/createConnection";
import { createServer } from "./server";

const startServer = async () => {
  const connection = await createConnection();
  const server = createServer(connection);

  const app = express();
  server.applyMiddleware({ app });

  const distDir = path.join(__dirname, "../../../web/dist");
  app.use(express.static(distDir));
  app.get("/*", (req, res) => {
    res.sendFile(path.join(distDir, "index.html"));
  });

  app.listen({ port: PORT });

  return server;
};

startServer()
  .then((server) => {
    console.log(
      `ðŸš€ Server ready at http://localhost:${PORT}${server.graphqlPath}`
    );
  })
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
