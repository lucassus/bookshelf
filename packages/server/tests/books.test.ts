import { ApolloServer, gql } from "apollo-server-express";
import { createTestClient } from "apollo-server-testing";
import { getManager } from "typeorm";

import { Book } from "../src/database/entity/Book";
import {
  createAuthor,
  createBook,
  createBookCopy,
  createUser
} from "../src/database/factories";
import { secureId } from "../src/database/helpers";
import { createServer } from "../src/server";

let server: ApolloServer;

beforeEach(async () => {
  server = createServer();
});

it("fetches books", async () => {
  // Given
  const { query } = createTestClient(server);

  const book = await createBook({ title: "Hobbit" });
  await createBookCopy({ bookId: book.id, ownerAttributes: { name: "John" } });
  const borrower = await createUser({ name: "Paul" });
  await createBookCopy({ bookId: book.id, borrowerId: borrower.id });

  const author = await createAuthor({ name: "George Lucas" });
  await createBook({ authorId: author.id, title: "Star Wars IV" });
  await createBook({ authorId: author.id, title: "Star Wars V" });

  // When
  const res = await query({
    query: gql`
      query {
        booksCount
        books {
          id
          title
          cover {
            url
          }
          copies {
            owner {
              id
              name
            }
            borrower {
              id
              name
            }
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();
});

it("fetches books with authors", async () => {
  // Given
  const { query } = createTestClient(server);

  const author = await createAuthor({ name: "Tolkien" });
  await createBook({ title: "Hobbit", authorId: author.id });
  await createBook({ title: "Lord of the Rings", authorId: author.id });
  await createBook();
  await createBook();

  // When
  const res = await query({
    query: gql`
      query {
        booksCount
        books {
          id
          title
          author {
            id
            name
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();
});

it("fetches a book", async () => {
  // Given
  const { query } = createTestClient(server);

  const author = await createAuthor({ name: "Andrzej Sapkpwski" });
  const book = await createBook({
    title: "Blood of Elves",
    authorId: author.id
  });

  const owner = await createUser({ name: "John" });
  const borrower = await createUser({ name: "Paul" });
  await createBookCopy({
    bookId: book.id,
    ownerId: owner.id,
    borrowerId: borrower.id
  });
  await createBookCopy({ bookId: book.id });

  // When
  const res = await query({
    query: gql`
      query GetBook($id: ID!) {
        book(id: $id) {
          id
          title
          description
          copies {
            id
            owner {
              id
              name
              avatar {
                image {
                  url
                }
                color
              }
            }
            borrower {
              id
              name
              avatar {
                image {
                  url
                }
                color
              }
            }
          }
        }
      }
    `,
    variables: { id: secureId.toExternal(book.id, "Book") }
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();
});

it("fetches a book with details", async () => {
  // Given
  const { query } = createTestClient(server);

  const book = await createBook({
    title: "The lady of the lake",
    authorAttributes: {
      name: "Andrzej Sapkowski"
    }
  });

  const GetBookWithDetailsQuery = gql`
    query GetBookWithDetails(
      $id: ID!
      $includeDetails: Boolean!
      $hideCover: Boolean = true
    ) {
      book(id: $id) {
        id
        title
        description @include(if: $includeDetails)
        author @include(if: $includeDetails) {
          name
        }

        cover @skip(if: $hideCover) {
          url
        }
      }
    }
  `;

  // When
  let res = await query({
    query: GetBookWithDetailsQuery,
    variables: {
      id: secureId.toExternal(book.id, "Book"),
      includeDetails: true,
      hideCover: false
    }
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();

  // When
  res = await query({
    query: GetBookWithDetailsQuery,
    variables: {
      id: secureId.toExternal(book.id, "Book"),
      includeDetails: false
    }
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();
});

it("responds with error when book cannot be found", async () => {
  // Given
  const { query } = createTestClient(server);

  // When
  const res = await query({
    query: gql`
      query GetBook($id: ID!) {
        book(id: $id) {
          id
          title
        }
      }
    `,
    variables: { id: secureId.toExternal(200, "Book") }
  });

  // Then
  expect(res.data).toBe(null);
  expect(res.errors![0].message).toEqual(
    'Could not find any entity of type "Book" matching: "200"'
  );
});

it("fetches books along with authors and books again", async () => {
  // Given
  const { query } = createTestClient(server);

  const author = await createAuthor({ name: "Andrzej Sapkowski" });
  await createBook({ authorId: author.id });
  await createBook({ authorId: author.id });

  // When
  const res = await query({
    query: gql`
      query {
        books {
          title
          author {
            name
            books {
              title
            }
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();
});

it("fetches a random book", async () => {
  // Given
  const { query } = createTestClient(server);

  const author = await createAuthor({ name: "Andrzej Sapkowski" });
  await createBook({
    authorId: author.id,
    title: "The tower of the swallow",
    coverPath: "/images/book-covers/witcher4.jpg"
  });

  // When
  const res = await query({
    query: gql`
      query {
        randomBook {
          title
          cover {
            url
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data!.randomBook).toMatchSnapshot();
});

it("updates book favourite", async () => {
  // Given
  const { mutate } = createTestClient(server);

  const book = await createBook({
    title: "Harry Potter and the Sorcerer's Stone",
    favourite: false
  });

  // When
  const res = await mutate({
    mutation: gql`
      mutation UpdateBookFavourite($id: ID!, $favourite: Boolean!) {
        updateBookFavourite(id: $id, favourite: $favourite) {
          id
          title
          favourite
        }
      }
    `,
    variables: { id: secureId.toExternal(book.id, "Book"), favourite: true }
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data!.updateBookFavourite).toMatchSnapshot();

  const updatedBook = await getManager().findOneOrFail(Book, 1);
  expect(updatedBook.favourite).toBe(true);
});
