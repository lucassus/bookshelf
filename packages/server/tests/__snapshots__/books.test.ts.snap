// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches a book 1`] = `
Object {
  "book": Object {
    "copies": Array [
      Object {
        "borrower": Object {
          "id": "MS1Vc2Vy",
          "name": "Alice",
        },
        "owner": Object {
          "id": "Mi1Vc2Vy",
          "name": "Bob",
        },
      },
    ],
    "description": "For more than a hundred years, humans, dwarves, gnomes and elves lived together in relative peace. But times have changed, the uneasy peace is over and now the races are fighting once again - killing their own kind and each other.

      Into this tumultuous time is born a child of prophecy, Ciri, surviving heiress of a bloody revolution, whose strange abilities can change the world - for good, or for evil...
      
      As the threat of war hangs over the land, Geralt the Witcher must protect Ciri from those who are hunting the child for her destructive power.
      
      But this time, Geralt may have met his match.",
    "id": "MTYtQm9vaw==",
    "title": "Blood of Elves",
  },
}
`;

exports[`fetches a book 2 1`] = `
Object {
  "book": Object {
    "copies": Array [
      Object {
        "borrower": null,
        "owner": Object {
          "id": "Mi1Vc2Vy",
          "name": "Bob",
        },
      },
    ],
    "description": "After walking through a portal in the Tower of the Swallow, thus narrowly escaping death, the Witcher girl, Ciri, finds herself in a completely different world... a world of the Elves. She is trapped with no way out. Time does not seem to exist and there are no obvious borders or portals to cross back into her home world.

      But this is Ciri, the child of prophecy, and she will not be defeated. She knows she must escape to finally rejoin the Witcher, Geralt, and his companions - and also to try to conquer her worst nightmare. Leo Bonhart, the man who chased, wounded and tortured Ciri, is still on her trail.
      
      And the world is still at war.",
    "id": "MjAtQm9vaw==",
    "title": "The lady of the lake",
  },
}
`;

exports[`fetches a random book 1`] = `
Object {
  "cover": Object {
    "url": "http://examples.devmastery.pl/assets/images/book-covers/witcher4.jpg",
  },
  "title": "The tower of the swallow",
}
`;

exports[`fetches books 1`] = `
Object {
  "books": Array [
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/expanse3.jpg",
      },
      "id": "MTAtQm9vaw==",
      "title": "Abaddon's Gate",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/expanse6.jpg",
      },
      "id": "MTMtQm9vaw==",
      "title": "Babylon's Ashes",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/witcher3.jpg",
      },
      "id": "MTgtQm9vaw==",
      "title": "Baptism of fire",
    },
    Object {
      "copies": Array [
        Object {
          "borrower": Object {
            "id": "MS1Vc2Vy",
            "name": "Alice",
          },
          "owner": Object {
            "id": "Mi1Vc2Vy",
            "name": "Bob",
          },
        },
      ],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/witcher1.jpg",
      },
      "id": "MTYtQm9vaw==",
      "title": "Blood of Elves",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/expanse2.jpg",
      },
      "id": "OS1Cb29r",
      "title": "Caliban's War",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/expanse4.jpg",
      },
      "id": "MTEtQm9vaw==",
      "title": "Cibola Burn",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/harry2.jpg",
      },
      "id": "Mi1Cb29r",
      "title": "Harry Potter and the Chamber of Secrets",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/harry7.jpg",
      },
      "id": "Ny1Cb29r",
      "title": "Harry Potter and the Deathly Hallows",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "url": "http://examples.devmastery.pl/assets/images/book-covers/harry4.jpg",
      },
      "id": "NC1Cb29r",
      "title": "Harry Potter and the Goblet of Fire",
    },
  ],
  "booksCount": 20,
}
`;

exports[`fetches books along with authors and books again 1`] = `
Object {
  "books": Array [
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Abaddon's Gate",
          },
          Object {
            "title": "Babylon's Ashes",
          },
          Object {
            "title": "Caliban's War",
          },
          Object {
            "title": "Cibola Burn",
          },
          Object {
            "title": "Leviathan Wakes",
          },
          Object {
            "title": "Nemesis Games",
          },
          Object {
            "title": "Persepolis Rising",
          },
          Object {
            "title": "Tiamat's Wrath",
          },
        ],
        "name": "James S. A. Corey",
      },
      "title": "Abaddon's Gate",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Abaddon's Gate",
          },
          Object {
            "title": "Babylon's Ashes",
          },
          Object {
            "title": "Caliban's War",
          },
          Object {
            "title": "Cibola Burn",
          },
          Object {
            "title": "Leviathan Wakes",
          },
          Object {
            "title": "Nemesis Games",
          },
          Object {
            "title": "Persepolis Rising",
          },
          Object {
            "title": "Tiamat's Wrath",
          },
        ],
        "name": "James S. A. Corey",
      },
      "title": "Babylon's Ashes",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Baptism of fire",
          },
          Object {
            "title": "Blood of Elves",
          },
          Object {
            "title": "The lady of the lake",
          },
          Object {
            "title": "The tower of the swallow",
          },
          Object {
            "title": "Time of contempt",
          },
        ],
        "name": "Andrzej Sapkowski",
      },
      "title": "Baptism of fire",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Baptism of fire",
          },
          Object {
            "title": "Blood of Elves",
          },
          Object {
            "title": "The lady of the lake",
          },
          Object {
            "title": "The tower of the swallow",
          },
          Object {
            "title": "Time of contempt",
          },
        ],
        "name": "Andrzej Sapkowski",
      },
      "title": "Blood of Elves",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Abaddon's Gate",
          },
          Object {
            "title": "Babylon's Ashes",
          },
          Object {
            "title": "Caliban's War",
          },
          Object {
            "title": "Cibola Burn",
          },
          Object {
            "title": "Leviathan Wakes",
          },
          Object {
            "title": "Nemesis Games",
          },
          Object {
            "title": "Persepolis Rising",
          },
          Object {
            "title": "Tiamat's Wrath",
          },
        ],
        "name": "James S. A. Corey",
      },
      "title": "Caliban's War",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Abaddon's Gate",
          },
          Object {
            "title": "Babylon's Ashes",
          },
          Object {
            "title": "Caliban's War",
          },
          Object {
            "title": "Cibola Burn",
          },
          Object {
            "title": "Leviathan Wakes",
          },
          Object {
            "title": "Nemesis Games",
          },
          Object {
            "title": "Persepolis Rising",
          },
          Object {
            "title": "Tiamat's Wrath",
          },
        ],
        "name": "James S. A. Corey",
      },
      "title": "Cibola Burn",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Harry Potter and the Chamber of Secrets",
          },
          Object {
            "title": "Harry Potter and the Deathly Hallows",
          },
          Object {
            "title": "Harry Potter and the Goblet of Fire",
          },
          Object {
            "title": "Harry Potter and the Half-Blood Prince",
          },
          Object {
            "title": "Harry Potter and the Order of the Phoenix",
          },
          Object {
            "title": "Harry Potter and the Prisoner of Azkaban",
          },
          Object {
            "title": "Harry Potter and the Sorcerer's Stone",
          },
        ],
        "name": "J. K. Rowling",
      },
      "title": "Harry Potter and the Chamber of Secrets",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Harry Potter and the Chamber of Secrets",
          },
          Object {
            "title": "Harry Potter and the Deathly Hallows",
          },
          Object {
            "title": "Harry Potter and the Goblet of Fire",
          },
          Object {
            "title": "Harry Potter and the Half-Blood Prince",
          },
          Object {
            "title": "Harry Potter and the Order of the Phoenix",
          },
          Object {
            "title": "Harry Potter and the Prisoner of Azkaban",
          },
          Object {
            "title": "Harry Potter and the Sorcerer's Stone",
          },
        ],
        "name": "J. K. Rowling",
      },
      "title": "Harry Potter and the Deathly Hallows",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Harry Potter and the Chamber of Secrets",
          },
          Object {
            "title": "Harry Potter and the Deathly Hallows",
          },
          Object {
            "title": "Harry Potter and the Goblet of Fire",
          },
          Object {
            "title": "Harry Potter and the Half-Blood Prince",
          },
          Object {
            "title": "Harry Potter and the Order of the Phoenix",
          },
          Object {
            "title": "Harry Potter and the Prisoner of Azkaban",
          },
          Object {
            "title": "Harry Potter and the Sorcerer's Stone",
          },
        ],
        "name": "J. K. Rowling",
      },
      "title": "Harry Potter and the Goblet of Fire",
    },
  ],
}
`;

exports[`updates book favourite 1`] = `
Object {
  "favourite": true,
  "id": "MS1Cb29r",
  "title": "Harry Potter and the Sorcerer's Stone",
}
`;
